src/heartbeats.cpp

```diff
@@ -0,0 +1,605 @@
+        }
+        else
+        {
+            LOG_warn << "discarding heartbeat update for pending sync: no valid id received for sync: " << pendingSyncPair->second->getLocalFolder();
+
+        }
+        mPendingSyncUpdates.erase(pendingSyncPair);
+    }
+}
+
+int MegaBackupMonitor::getHBState(MegaSync *sync)
+{
+    if (sync->isTemporaryDisabled())
+    {
+        return MegaBackupMonitor::State::TEMPORARY_DISABLED;
+    }
+    else if (sync->isActive())
+    {
+        //TODO: consider use case: paused, if transfers are paused!?
+        return MegaBackupMonitor::State::ACTIVE;
+    }
+    else if (!sync->isEnabled())
+    {
+        return MegaBackupMonitor::State::DISABLED;
+    }
+    else
+    {
+        return MegaBackupMonitor::State::FAILED;
+    }
+}
+
+int MegaBackupMonitor::getHBSubstatus(MegaSync *sync)
+{
+    return sync->getError();
+}
+
+string MegaBackupMonitor::getHBExtraData(MegaSync *sync)
+{
+    return "";
+}
+
+BackupType MegaBackupMonitor::getHBType(MegaSync *sync)
+{
+    return BackupType::TWO_WAY; // TODO: get that from sync whenever others are supported
+}
+
+void MegaBackupMonitor::updateSyncInfo(handle backupId, MegaSync *sync)
+{
+
+    string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", sync->getLocalFolder()) );
+    string deviceIDEncrypted(mClient->cypherTLVTextWithMasterKey("de", mClient->getDeviceid()) );
+    string nameEncrypted(mClient->cypherTLVTextWithMasterKey("na", sync->getName()) );
+
+
+    mClient->reqs.add(new CommandBackupPut(mClient, backupId, BackupType::TWO_WAY, sync->getMegaHandle(), localFolderEncrypted.c_str(),
+                                           deviceIDEncrypted.c_str(), nameEncrypted.c_str(),
+                                           getHBState(sync), getHBSubstatus(sync), getHBExtraData(sync).c_str()
+                                           ));
+}
+
+void MegaBackupMonitor::updateOrRegisterSync(MegaSync *sync)
+{
+    if (!sync)
+    {
+        return;
+    }
+    auto config = mClient->syncConfigs->get(sync->getTag());
+
+    handle syncID = UNDEF;
+    if (config)
+    {
+        syncID = config->getBackupId();
+    }
+
+    bool pushInPending = false;
+
+    string extraData;
+    if (syncID == UNDEF) //register
+    {
+
+        string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", sync->getLocalFolder()) );
+        string deviceIDEncrypted(mClient->cypherTLVTextWithMasterKey("de", mClient->getDeviceid()) );
+        string nameEncrypted(mClient->cypherTLVTextWithMasterKey("na", sync->getName()) );
+
+
+        if (std::find(mPendingBackupPuts.begin(), mPendingBackupPuts.end(), sync->getTag()) == mPendingBackupPuts.end())
+        {
+            pushInPending = true;
+            mClient->reqs.add(new CommandBackupPut(mClient, BackupType::TWO_WAY, sync->getMegaHandle(), localFolderEncrypted.c_str(),
+                                           deviceIDEncrypted.c_str(), nameEncrypted.c_str(),
```

Comment:
We should avoid to create a new string (this signature of the command's ctor receives const references to string directly). I'll fix it.

Reply:
Done!